### PR TITLE
CollectDataToIoRank: don't require the elements of a dune grid to be traversed in consecutive order

### PR DESCRIPTION
**Pull request by andlaus**
I'm sure that the results for sequential runs will be identical, but I'm not sure if the last three hunks are semantically correct.

### CODE REVIEW

File: ebos/collecttoiorank.hh [hunk 1]
```diff
@@ -193,22 +193,19 @@ namespace Ewoms
...
 #if DUNE_VERSION_NEWER(DUNE_COMMON, 2,4)
                         int elemIdx = elemMapper.index(*elemIt );
 #else
                         int elemIdx = elemMapper.map(*elemIt );
 #endif
-                        assert( count == elemIdx );
                         int cartElemIdx = gridManager.equilCartesianIndexMapper().cartesianIndex(elemIdx);
-                        globalCartesianIndex_.push_back( cartElemIdx );
+                        globalCartesianIndex_[elemIdx] = cartElemIdx;
                     }
```
Comment (dr-robertk) [inline ebos/collecttoiorank.hh]: Looks ok.

**Comment by dr-robertk** [comment]
Looks good. I'll merge.

File: ebos/collecttoiorank.hh [hunk 2]
```diff
@@ -248,7 +244,7 @@ namespace Ewoms
                     // only store interior element for collection
                     if( element.partitionType() == Dune :: InteriorEntity )
                     {
-                        localIndexMap_.push_back( index );
+                        localIndexMap_[elemIdx] = elemIdx;
```
Comment (blattms) [inline ebos/collecttoiorank.hh]: Did anybody test/run this code before merging? To the very least this looks suspicious (why would one need to store an identity map?) and seems to be a buggy change of semantics. Note that only interior entities were pushed previously (resulting in a map of entity index (by traversal) to interior entity index (by traversal). BTW Is the size of localIndexMap_ ever set?

Comment (andlaus) [inline ebos/collecttoiorank.hh]: I ran it, albeit not parallel. I can't judge if that array is required, but I'm pretty sure that the result is the same as before if `Dune::CpGrid` is used for the grid.

**Comment by blattms** [comment]
> I ran it, albeit not parallel. I can't judge if that array is required, but I'm pretty sure that the result is the same as before if `Dune::CpGrid` is used for the grid.

Look again, in a parallel run you will have non-interior entities.

**Comment by andlaus** [comment]
that code is _really_ non-obvious IMO.

**Pull request merged**